### PR TITLE
Py3update

### PR DESCRIPTION
**Pull request by atfienberg**
Initial updates for python3 migration

### CODE REVIEW

File: pisa/core/container.py [hunk 1]
```diff
@@ -417,7 +417,7 @@ def binned_to_array(self, key):
         try:
             binning, hist = self.binned_data[key]
         except KeyError:
-            if self.array_data.has_key(key):
+            if self.key in array_data:
```
Comment (jllanfranchi) [inline pisa/core/container.py]: fix

Comment (atfienberg) [reply pisa/core/container.py]: done

**Comment by jllanfranchi** [comment]
Still references to `basestring` found in the following files:
```pylintrc
pisa/core/param.py
pisa/utils/PIDSpec.py
pisa/utils/mcSimRunSettings.py
pisa/stages/aeff/hist.py
pisa/stages/aeff/smooth.py
pisa/stages/reco/hist.py
pisa/stages/reco/simple_param.py
pisa/stages/aeff/param.py
pisa/stages/xsec/genie.py
pisa/stages/reco/vbwkde.py
pisa/stages/reco/param.py
pisa/stages/pid/hist.py
pisa/stages/unfold/roounfold.py
pisa/stages/pid/smooth.py
pisa/stages/pid/param.py
pisa/stages/data/events_to_data.py
pisa/stages/flux/mceq.py
pisa/stages/discr_sys/fit.py
pisa/stages/data/sample.py
```

**Comment by jllanfranchi** [comment]
You broke `utils/cache.py` (execute the file, produces recursion errors for me):

```$ python ~/src/pisa/pisa/utils/cache.py
<< PISA is running in double precision (FP64) mode; numba is running on GPU >>
Traceback (most recent call last):
  File "/home/justin/src/pisa/pisa/utils/cache.py", line 532, in <module>
    test_MemoryCache()
  File "/home/justin/src/pisa/pisa/utils/cache.py", line 484, in test_MemoryCache
    assert 0 not in mc
  File "/home/justin/src/pisa/pisa/utils/cache.py", line 124, in __contains__
    return key in self
  File "/home/justin/src/pisa/pisa/utils/cache.py", line 124, in __contains__
    return key in self
  File "/home/justin/src/pisa/pisa/utils/cache.py", line 124, in __contains__
    return key in self
  [Previous line repeated 995 more times]
RecursionError: maximum recursion depth exceeded
```

**Comment by philippeller** [comment]
I would recommend to merge this PR soon. So far we have not heard of big issues and should we encounter any in the future the only real option is to fix them and make it work with py3 anyways. Any objections?

**Pull request merged**